tfhe/src/integer/server_key/radix/shift.rs

```diff
@@ -117,50 +117,82 @@ impl ServerKey {
+        let shift_within_block = shift % num_bits_in_block;
+        let num_blocks = ct.blocks.len();
 
-            // We partially propagate in order to not lose information
-            self.partial_propagate(ct);
-            self.blockshift_right_assign(ct, 1_usize);
+        // rotate left as the blocks are from LSB to MSB
+        ct.blocks.rotate_left(rotations);
+        for block in &mut ct.blocks[num_blocks - rotations..] {
+            *block = self.key.create_trivial(0);
+        }
 
-            // We propagate the last block in order to not lose information
-            self.propagate(ct, ct.blocks.len() - 2);
-            self.blockshift_right_assign(ct, quotient);
+        if shift_within_block == 0 || rotations == ct.blocks.len() {
+            return;
         }
-    }
 
-    /// Propagates all carries except the last one.
-    /// For development purpose only.
-    fn partial_propagate<PBSOrder: PBSOrderMarker>(&self, ctxt: &mut RadixCiphertext<PBSOrder>) {
-        let len = ctxt.blocks.len() - 1;
-        for i in 0..len {
-            self.propagate(ctxt, i);
+        let message_modulus = self.key.message_modulus.0 as u64;
+        let lut =
+            self.key
+                .generate_accumulator_bivariate(|mut current_block, mut previous_block| {
+                    // left shift not to lose
+                    // bits when shifting right afterwards
+                    current_block <<= num_bits_in_block;
+                    previous_block <<= num_bits_in_block;
+
+                    current_block >>= shift_within_block;
+                    previous_block >>= shift_within_block;
+
+                    // The way of gettint carry / message is reversed compared
+                    // to the usual way but its normal
+                    let message_of_current_block = current_block >> num_bits_in_block;
+                    let carry_of_previous_block = previous_block % message_modulus;
+
+                    message_of_current_block + carry_of_previous_block
+                });
+        let partial_blocks = ct.blocks[..num_blocks - rotations]
+            .windows(2)
+            .map(|blocks| {
+                // We are right-shifting, which in our representation
+                // means the previous_block (the one with the carry to be
+                // progatated to the current_block) is the next block in the Vec
+                let (current_block, previous_block) = (&blocks[0], &blocks[1]);
+                self.key
+                    .unchecked_apply_lookup_table_bivariate(current_block, previous_block, &lut)
+            })
+            .collect::<Vec<_>>();
+        self.key.unchecked_scalar_right_shift_assign(
+            &mut ct.blocks[num_blocks - rotations - 1],
+            shift_within_block as u8,
+        );
+        self.key
+            .message_extract_assign(&mut ct.blocks[num_blocks - rotations - 1]);
+        assert_eq!(partial_blocks.len(), ct.blocks.len() - rotations - 1);
```

Comment:
could we maybe create the mutable slice first and then compare its length to the partial_blocks len ? Would make the comparison a bit less confusing, or maybe use the izip! macro from core crypto I think it does some bound check (or zip checked) but I'm sure we have a zip that verifies iterators have the same len